library/manageiq_custom_attributes.py

```diff
@@ -0,0 +1,259 @@
+  miq_password:
+    description:
+      - manageiq password
+    default: MIQ_PASSWORD env var if set. otherwise, it is required to pass it
+  entity_name:
+    description:
+      - the entity name in manageiq to which the custom attributes belongs
+    required: true
+    default: null
+  entity_type:
+    description:
+      - the entity type in manageiq to which the custom attributes belongs
+    required: true
+    default: null
+  state:
+    description:
+      - the state of the custom attributes
+      - On present, it will add the custom attributes to the entity if not
+      already exist, or update the custom attributes if the associated data
+      is different
+      - On absent, it will delete the custom attributes from the entity if exist
+    required: false
+    choices: ['present', 'absent']
+    default: 'present'
+  custom_attributes:
+    description:
+      - the custom attributes of the entity
+    required: true
+    default: null
+'''
+
+EXAMPLES = '''
+# Add custom attributes for Openshift Containers Provider in ManageIQ
+  manageiq_custom_attributes:
+    entity_type: 'provider'
+    entity_name: 'openshift01'
+    state: 'present'
+    miq_url: 'http://localhost:3000'
+    miq_username: 'admin'
+    miq_password: '******'
+    custom_attributes:
+      - name: "ca1"
+        value: "value 1"
+      - name: "ca2"
+        value: "value 2"
+'''
+
+
+class ManageIQCustomAttributes(object):
+    """ ManageIQ object to execute custom attibutes related operations
+    in manageiq
+
+    url      - manageiq environment url
+    user     - the username in manageiq
+    password - the user password in manageiq
+    """
+
+    supported_entities = {'vm': 'vms', 'provider': 'providers'}
+
+    def __init__(self, module, url, user, password):
+        self.module        = module
+        self.api_url       = url + '/api'
+        self.user          = user
+        self.password      = password
+        self.client        = MiqApi(self.api_url, (self.user, self.password))
+        self.changed       = False
+
+    def find_entity_by_name(self, entity_type, entity_name):
+        """ Searches the entity name in ManageIQ.
+
+            Returns:
+                the entity id if it exists in manageiq, None otherwise.
+        """
+        entities_list = getattr(self.client.collections, ManageIQCustomAttributes.supported_entities[entity_type])
+        return next((e.id for e in entities_list if e.name == entity_name), None)
+
+    def get_entity_cas(self, entity_type, entity_id):
+        """ Returns the entity's custom attributes
+        """
+        try:
+            url = '{api_url}/{entity_type}/{id}?expand=custom_attributes'.format(
+                   api_url=self.api_url,
+                   entity_type=ManageIQCustomAttributes.supported_entities[entity_type],
+                   id=entity_id)
+            result = self.client.get(url)
+            return result.get('custom_attributes')
```

Comment:
> Can API omit this key despite ?expand=custom_attributes?

Yes.

> if it simply means there were no attributes, should return [].

you mean instead of returning None? sure, I'll make the change